src/python/pants/source/source_root.py

```diff
@@ -202,31 +222,59 @@ class OptionalSourceRoot:
 
 
 @rule
-def get_source_root(
+async def get_source_root(
     source_root_request: SourceRootRequest, source_root_config: SourceRootConfig
 ) -> OptionalSourceRoot:
     """Rule to request a SourceRoot that may not exist."""
-    matched_path = source_root_config.get_pattern_matcher().find_root(source_root_request.path)
-    if matched_path:
-        return OptionalSourceRoot(SourceRoot(path=str(matched_path)))
-    else:
-        return OptionalSourceRoot(None)
+    if ".." in str(source_root_request):
+        raise NoSourceRootError(
+            str(source_root_request), "`..` disallowed in source root searches. "
```

Comment:
This should probably move into the `SourceRootRequest` class. See #9538 for some other code that does this that might be worth extracting to more generally check for normalization.

Reply:
Good idea.